lib/new_relic/traced_thread.rb

```diff
@@ -0,0 +1,46 @@
+# encoding: utf-8
+# This file is distributed under New Relic's license terms.
+# See https://github.com/newrelic/newrelic-ruby-agent/blob/main/LICENSE for complete details.
+# frozen_string_literal: true
+
+module NewRelic
+  #
+  # This class allows the state of a transaction to be passed to a Thread so that nested segments can be created from the operations performed within the Thread's block.
+  # To have the New Relic Ruby agent automatically trace all of your applications threads,
+  # enable the `instrumentation.thread.tracing` configuration option in your newrelic.yml
+  #
+  # Note: disabling the configuration option `instrumentation.thread` while using this class can cause incorrectly nested spans
+  #
+  # @api public
+  class TracedThread < Thread
+    #
+    # Creates a new Thread whose work will be traced by New Relic.
+    # Use this class as a replacement for the native Thread class.
+    # Example: instead of using `Thread.new`, use
+    # ```ruby
+    #   NewRelic::TracedThread.new { execute_some_code }
+    # ```
+    #
+    # @api public
+    def initialize(*args, &block)
+      NewRelic::Agent.record_api_supportability_metric(:traced_thread)
+      traced_block = create_traced_block(*args, &block)
+      super(*args, &traced_block)
+    end
+
+    def create_traced_block(*args, &block)
+      return block if NewRelic::Agent.config[:'instrumentation.thread.tracing'] # if this is on, don't double trace
+
+      instrumentation = ::Thread.current[:newrelic_tracer_state]
+      Proc.new do
+        begin
+          ::Thread.current[:newrelic_tracer_state] = instrumentation
+          segment = NewRelic::Agent::Tracer.start_segment(name: "Thread#{::Thread.current.object_id}")
+          block.call(*args) if block.respond_to?(:call)
+        ensure
+          segment.finish if segment
+        end
```

Comment:
Since these lines duplicate the code in `lib/new_relic/agent/instrumentation/thread/instrumentation.rb`, I wonder if a separate Helper/Concern/Class could be created that held the logic for both? This might help avoid problems down the line if/when this code needs to be updated so that it can be maintained in one location.

Reply:
Yeah that's a good idea. I ended up putting the new method inside of Tracer, since that is where the other methods that deal with the state live.